File: src/MphRead.Native/Mods/ScreenCapture.cpp
#include "ScreenCapture.hpp"

#include "../NativeRuntime/OpenTK/GL.hpp"
#include "../NativeRuntime/System/Console.hpp"
#include "../NativeRuntime/System/Globalization.hpp"

#include <typeinfo>

#include "ThumbnailLog.hpp"
#include "../Formats/Types.hpp"
#include "../Scene.hpp"
#include "../NativeRuntime/System/ExceptionText.hpp"
#include "../NativeRuntime/System/IO.hpp"
#include "../NativeRuntime/System/Managed.hpp"

#include <bit>
#include <cstdint>
#include <exception>
#include <filesystem>
#include <fstream>
#include <optional>
#include <ostream>
#include <span>
#include <string>
#include <string_view>
#include <utility>
#include <vector>

using ::MphRead::NativeRuntime::ExceptionTypeName;
using ::MphRead::NativeRuntime::PathFromUtf8;
using ::MphRead::NativeRuntime::PathToUtf8;
using ::MphRead::NativeRuntime::UncheckedAdd;
using ::MphRead::NativeRuntime::UncheckedMultiply;

namespace MphRead::Export::ImagesInterop
{
    void SetFlipVerticallyOnSave(bool value);
    void WritePngRgb(
        std::span<const std::uint8_t> buffer,
        std::int32_t width,
        std::int32_t height,
        std::ostream& stream);
}

namespace
{
    constexpr std::int32_t GlDebugOutput = 0x92E0;
    constexpr std::int32_t GlDebugOutputSynchronous = 0x8242;
    constexpr std::int32_t GlDebugSeverityNotification = 0x826B;
    constexpr std::int32_t GlVendor = 0x1F00;
    constexpr std::int32_t GlRenderer = 0x1F01;
    constexpr std::int32_t GlVersion = 0x1F02;
    constexpr std::int32_t GlContextFlags = 0x821E;
    constexpr std::int32_t GlContextProfileMask = 0x9126;
    constexpr std::int32_t GlContextFlagForwardCompatibleBit = 0x00000001;
    constexpr std::int32_t GlContextCoreProfileBit = 0x00000001;
    constexpr std::int32_t GlContextCompatibilityProfileBit = 0x00000002;

    [[nodiscard]] std::ofstream CreateFile(const std::string& path)
    {
        std::ofstream stream(PathFromUtf8(path), std::ios::binary | std::ios::trunc);
        stream.exceptions(std::ios::failbit | std::ios::badbit);
        return stream;
    }

    [[nodiscard]] std::string DebugSourceName(std::int32_t value)
    {
        switch (value)
        {
        case 0x8246: return "DebugSourceApi";
        case 0x8247: return "DebugSourceWindowSystem";
        case 0x8248: return "DebugSourceShaderCompiler";
        case 0x8249: return "DebugSourceThirdParty";
        case 0x824A: return "DebugSourceApplication";
        case 0x824B: return "DebugSourceOther";
        default: return std::to_string(value);
        }
    }

    [[nodiscard]] std::string DebugTypeName(std::int32_t value)
    {
        switch (value)
        {
        case 0x824C: return "DebugTypeError";
        case 0x824D: return "DebugTypeDeprecatedBehavior";
        case 0x824E: return "DebugTypeUndefinedBehavior";
        case 0x824F: return "DebugTypePortability";
        case 0x8250: return "DebugTypePerformance";
        case 0x8251: return "DebugTypeOther";
        case 0x8268: return "DebugTypeMarker";
        case 0x8269: return "DebugTypePushGroup";
        case 0x826A: return "DebugTypePopGroup";
        default: return std::to_string(value);
        }
    }

    [[nodiscard]] std::string DebugSeverityName(std::int32_t value)
    {
        switch (value)
        {
        case 0x9146: return "DebugSeverityHigh";
        case 0x9147: return "DebugSeverityMedium";
        case 0x9148: return "DebugSeverityLow";
        case 0x826B: return "DebugSeverityNotification";
        default: return std::to_string(value);
        }
    }

    void InvokeReport(
        const std::function<void(const std::string&)>& report,
        const std::string& line)
    {
        if (!report)
        {
            throw System::NullReferenceException();
        }
        report(line);
    }
}

namespace
{
    // GL_KHR_debug and the two glGet forms this file reads, which the managed
    // build reaches through OpenTK's own GL class.
    using GlDebugProc = void (*)(
        std::int32_t, std::int32_t, std::int32_t, std::int32_t, std::int32_t,
        const char*, const void*);

    void GlEnable(std::int32_t capability)
    {
        ::OpenTK::Graphics::OpenGL::GL::Enable(
            static_cast<::OpenTK::Graphics::OpenGL::GL::EnableCap>(capability));
    }

    void GlDebugMessageCallback(GlDebugProc callback, const void* userParam)
    {
        ::OpenTK::Graphics::OpenGL::GL::DebugMessageCallback(
            reinterpret_cast<void*>(callback), userParam);
    }

    [[nodiscard]] std::optional<std::string> GlGetString(std::int32_t name)
    {
        std::string value = ::OpenTK::Graphics::OpenGL::GL::GetString(
            static_cast<::OpenTK::Graphics::OpenGL::GL::StringName>(name));
        if (value.empty())
        {
            return std::nullopt;
        }
        return value;
    }

    [[nodiscard]] std::int32_t GlGetInteger(std::int32_t pname)
    {
        return ::OpenTK::Graphics::OpenGL::GL::GetInteger(pname);
    }

}

namespace MphRead::Mods
{
    ScreenCapture::PngWriterAction ScreenCapture::_pngWriter{};
    ScreenCapture::DebugProc ScreenCapture::_debugCallback{};
    std::int32_t ScreenCapture::_messagesLogged = 0;

    ScreenCapture::PngWriterAction ScreenCapture::PngWriter()
    {
        return _pngWriter;
    }

    void ScreenCapture::PngWriter(PngWriterAction value)
    {
        _pngWriter = std::move(value);
    }

    bool ScreenCapture::SaveWindow(Scene* scene, const std::string& path)
    {
        if (scene == nullptr)
        {
            throw System::NullReferenceException();
        }
        ReadPixels read = [scene](std::int32_t& width, std::int32_t& height)
        {
            return scene->ReadWindowBuffer(width, height);
        };
        return Save(scene, path, read);
    }

    bool ScreenCapture::Save(Scene* scene, const std::string& path)
    {
        if (scene == nullptr)
        {
            throw System::NullReferenceException();
        }
        ReadPixels read = [scene](std::int32_t& width, std::int32_t& height)
        {
            return scene->ReadSceneTarget(width, height);
        };
        return Save(scene, path, read);
    }

    bool ScreenCapture::Save(
        Scene* scene,
        const std::string& path,
        const ReadPixels& read)
    {
        (void)scene;
        try
        {
            std::int32_t width = 0;
            std::int32_t height = 0;
            PixelBuffer pixels = read(width, height);
            if (!pixels || width <= 0 || height <= 0)
            {
                return false;
            }

            if (LitFraction(*pixels) < MinLitFraction)
            {
                std::string first = PathToUtf8(PathFromUtf8(path).filename());
                first += " came out black ";

                std::string second = "(";
                second += NativeRuntime::DoubleToStringFixed2(
                    LitFraction(*pixels) * 100.0);
                second += "% lit, ";
                second += std::to_string(width);
                second += "x";
                second += std::to_string(height);
                second += "); not saving it. ";

                std::string third = "The scene rendered nothing -- ";
                third += DescribeContext();

                const std::string why = first + second + third;
                NativeRuntime::ConsoleWriteLine("[capture] " + why);
                ThumbnailLog::Write(why);
                return false;
            }

            const std::filesystem::path pathValue = PathFromUtf8(path);
            const std::filesystem::path directory
                = pathValue == pathValue.root_path()
                    ? std::filesystem::path{}
                    : pathValue.parent_path();
            if (!directory.empty())
            {
                std::filesystem::create_directories(directory);
            }

            PngWriterAction writer = PngWriter();
            if (writer)
            {
                writer(*pixels, width, height, path);
                return true;
            }

            std::ofstream stream = CreateFile(path);
            try
            {
                Export::ImagesInterop::SetFlipVerticallyOnSave(true);
                Export::ImagesInterop::WritePngRgb(
                    std::span<const std::uint8_t>(*pixels), width, height, stream);
            }
            catch (...)
            {
                try
                {
                    stream.close();
                }
                catch (...)
                {
                    throw;
                }
                throw;
            }
            stream.close();
            return true;
        }
        catch (const std::exception& exception)
        {
            NativeRuntime::ConsoleWriteLine(
                "[capture] could not save " + path + ": "
                    + std::string(exception.what()));
            return false;
        }
    }

    double ScreenCapture::LitFraction(const std::vector<std::uint8_t>& pixels)
    {
        const std::int32_t length = static_cast<std::int32_t>(pixels.size());
        const auto channel = [&](std::int32_t index) -> std::uint8_t
        {
            if (index < 0 || index >= length)
            {
                throw SceneDetail::IndexOutOfRangeException();
            }
            return pixels[static_cast<std::size_t>(index)];
        };

        std::int32_t lit = 0;
        std::int32_t total = 0;
        for (std::int32_t i = 0; UncheckedAdd(i, 2) < length; i = UncheckedAdd(i, 3))
        {
            total = UncheckedAdd(total, 1);
            if (channel(i) > 8
                || channel(UncheckedAdd(i, 1)) > 8
                || channel(UncheckedAdd(i, 2)) > 8)
            {
                lit = UncheckedAdd(lit, 1);
            }
        }
        return total == 0 ? 0.0 : static_cast<double>(lit) / static_cast<double>(total);
    }

    void ScreenCapture::DebugThunk(
        std::int32_t source,
        std::int32_t type,
        std::int32_t id,
        std::int32_t severity,
        std::int32_t length,
        const char* message,
        const void* param)
    {
        if (_debugCallback)
        {
            _debugCallback(source, type, id, severity, length, message, param);
        }
    }

    void ScreenCapture::EnableDebugOutput(ReportAction report)
    {
        try
        {
            _debugCallback = [report](
                std::int32_t source,
                std::int32_t type,
                std::int32_t id,
                std::int32_t severity,
                std::int32_t length,
                const char* message,
                const void* param)
            {
                (void)id;
                (void)param;
                if (severity == GlDebugSeverityNotification || _messagesLogged >= 12)
                {
                    return;
                }
                _messagesLogged = UncheckedAdd(_messagesLogged, 1);
                const std::string text
                    = std::string(message, static_cast<std::size_t>(length));
                const std::string severityText = DebugSeverityName(severity);
                const std::string typeText = DebugTypeName(type);
                const std::string sourceText = DebugSourceName(source);
                InvokeReport(
                    report,
                    "GL says: [" + severityText + "] " + typeText
                        + " from " + sourceText + ": " + text);
            };
            GlEnable(GlDebugOutput);
            GlEnable(GlDebugOutputSynchronous);
            GlDebugMessageCallback(&ScreenCapture::DebugThunk, nullptr);
        }
        catch (const std::exception& exception)
        {
            InvokeReport(
                report,
                "could not turn on GL debug output ("
                    + ExceptionTypeName(exception)
                    + "); this driver may not have KHR_debug");
        }
    }

    std::string ScreenCapture::DescribeContext()
    {
        try
        {
            const std::string vendor = GlGetString(GlVendor).value_or("?");
            const std::string renderer = GlGetString(GlRenderer).value_or("?");
            const std::string version = GlGetString(GlVersion).value_or("?");
            const std::int32_t flags = GlGetInteger(GlContextFlags);
            const std::string forward = (flags & GlContextFlagForwardCompatibleBit) != 0
                ? ", FORWARD-COMPATIBLE (deprecated entry points removed, which is all of immediate mode)"
                : "";
            const std::int32_t mask = GlGetInteger(GlContextProfileMask);
            const std::string profile = (mask & GlContextCoreProfileBit) != 0
                ? "CORE (immediate mode is unavailable, which renders everything black)"
                : (mask & GlContextCompatibilityProfileBit) != 0
                    ? "compatibility"
                    : "unreported";
            return "GL " + version + ", profile " + profile + forward
                + ", " + vendor + " / " + renderer;
        }
        catch (const std::exception& exception)
        {
            return "could not query the GL context: "
                + std::string(exception.what());
        }
    }

    double ScreenCapture::NonBlackFraction(Scene* scene)
    {
        if (scene == nullptr)
        {
            throw System::NullReferenceException();
        }

        std::int32_t width = 0;
        std::int32_t height = 0;
        PixelBuffer pixels = scene->ReadSceneTarget(width, height);
        if (!pixels || width <= 0)
        {
            return 0.0;
        }
        const std::int32_t length = static_cast<std::int32_t>(pixels->size());
        const auto channel = [&](std::int32_t index) -> std::uint8_t
        {
            if (index < 0 || index >= length)
            {
                throw SceneDetail::IndexOutOfRangeException();
            }
            return (*pixels)[static_cast<std::size_t>(index)];
        };

        std::int32_t lit = 0;
        for (std::int32_t i = 0; i < length; i = UncheckedAdd(i, 3))
        {
            if (channel(i) > 8
                || channel(UncheckedAdd(i, 1)) > 8
                || channel(UncheckedAdd(i, 2)) > 8)
            {
                lit = UncheckedAdd(lit, 1);
            }
        }
        const std::int32_t area = UncheckedMultiply(width, height);
        return static_cast<double>(lit) / static_cast<double>(area);
    }
}

File: src/MphRead/Mods/ScreenCapture.cs
using System;
using System.IO;
using OpenTK.Graphics.OpenGL;
using ReFuel.Stb;
using OpenTK.Windowing.GraphicsLibraryFramework;

namespace MphRead.Mods
{
    /// <summary>
    /// Saves what the scene just drew to a PNG.
    ///
    /// Reads the scene's own offscreen target rather than the window's back
    /// buffer. A hidden window has no usable back buffer under Mesa -- which
    /// is why headless captures came out black on Linux while the identical
    /// code produced correct thumbnails on Windows -- and even a visible one
    /// leaves the back buffer undefined once SwapBuffers has run. The
    /// offscreen target is a texture the scene owns, so it is valid either
    /// way.
    /// </summary>
    public static class ScreenCapture
    {
        /// <summary>
        /// How a PNG gets written: the pixels bottom-up in RGB, the width, the
        /// height and the path.
        ///
        /// STB everywhere but Android, where the package ships natives for
        /// Linux and Windows and none for a phone -- the room rendered and the
        /// frame read back fine, and then the encoder's type initializer threw.
        /// The head there sets this to Android's own encoder.
        /// </summary>
        public static Action<byte[], int, int, string>? PngWriter { get; set; }

        /// <summary>
        /// Save what the *window* holds, which is the only capture that
        /// includes the HUD. See <c>Scene.ReadWindowBuffer</c>: only valid on
        /// a visible window, and only between the draw and the buffer swap.
        /// </summary>
        public static bool SaveWindow(Scene scene, string path)
        {
            return Save(scene, path, scene.ReadWindowBuffer);
        }

        public static bool Save(Scene scene, string path)
        {
            return Save(scene, path, scene.ReadSceneTarget);
        }

        /// <summary>
        /// Save what the window holds when there is no scene at all -- the
        /// launcher, drawn into the game window with nothing behind it.
        ///
        /// Same rules as <see cref="SaveWindow(Scene, string)"/>: the window has to be
        /// visible, and this has to be called after the draw and before the
        /// buffer swap. There is no scene to describe if it comes out black,
        /// which is itself the answer -- a black frame here means the overlay
        /// drew nothing.
        /// </summary>
        public static bool SaveWindow(int width, int height, string path)
        {
            return Save(scene: null, path, (out int w, out int h) =>
            {
                w = width;
                h = height;
                if (width <= 0 || height <= 0)
                {
                    return null;
                }
                byte[] buffer = new byte[width * height * 3];
                GL.BindFramebuffer(FramebufferTarget.ReadFramebuffer, 0);
                GL.ReadBuffer(ReadBufferMode.Back);
                GL.PixelStore(PixelStoreParameter.PackAlignment, 1);
                GL.ReadPixels(0, 0, width, height, PixelFormat.Rgb, PixelType.UnsignedByte, buffer);
                return buffer;
            });
        }

        private delegate byte[]? ReadPixels(out int width, out int height);

        private static bool Save(Scene? scene, string path, ReadPixels read)
        {
            try
            {
                byte[]? pixels = read(out int width, out int height);
                if (pixels == null || width <= 0 || height <= 0)
                {
                    return false;
                }
                // An all-black frame is a failure that writes a file. It
                // looks like a success in every log, the caller counts it,
                // and the player ends up with a full set of black pictures
                // and nothing saying why -- which is exactly how it was
                // reported. Refusing to save it turns a silent wrong answer
                // into a loud missing one.
                if (LitFraction(pixels) < MinLitFraction)
                {
                    string why = $"{Path.GetFileName(path)} came out black "
                        + $"({LitFraction(pixels) * 100:0.00}% lit, {width}x{height}); not saving it. "
                        + $"The scene rendered nothing -- {DescribeContext()}";
                    Console.WriteLine($"[capture] {why}");
                    ThumbnailLog.Write(why);
                    return false;
                }
                string? directory = Path.GetDirectoryName(path);
                if (!string.IsNullOrEmpty(directory))
                {
                    Directory.CreateDirectory(directory);
                }
                Action<byte[], int, int, string>? writer = PngWriter;
                if (writer != null)
                {
                    writer(pixels, width, height, path);
                    return true;
                }
                using FileStream stream = File.Create(path);
                StbImage.FlipVerticallyOnSave = true;
                StbImage.WritePng<byte>(pixels, width, height, StbiImageFormat.Rgb, stream);
                return true;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"[capture] could not save {path}: {ex.Message}");
                return false;
            }
        }

        /// <summary>
        /// How much of a picture has to be lit before it counts as a picture.
        ///
        /// "Any lit pixel at all" is not enough, and was tried: a dead context
        /// still produces the odd non-zero pixel, so a retry eventually passed
        /// and the run reported a capture it had refused to save. A room seen
        /// from its own intro camera lights up far more than a hundredth of
        /// the frame.
        /// </summary>
        private const double MinLitFraction = 0.01;

        private static double LitFraction(byte[] pixels)
        {
            int lit = 0;
            int total = 0;
            for (int i = 0; i + 2 < pixels.Length; i += 3)
            {
                total++;
                if (pixels[i] > 8 || pixels[i + 1] > 8 || pixels[i + 2] > 8)
                {
                    lit++;
                }
            }
            return total == 0 ? 0 : lit / (double)total;
        }

        /// <summary>
        /// What the driver actually handed over, which is the one thing a
        /// black render never says on its own.
        ///
        /// This engine draws in immediate mode -- GL.Begin and friends -- and
        /// those do not exist in a core profile. A driver that answers a
        /// request for 3.2 Compatability with a core context therefore fails
        /// every draw call silently and renders black, with nothing in any log
        /// to say so. It is the documented cause on Mesa and it is not unique
        /// to Mesa.
        /// </summary>
        // Kept alive deliberately: the driver holds this pointer for the
        // lifetime of the context, and a delegate that is only a local goes
        // to the collector and takes the process with it on the next message.
        private static DebugProc? _debugCallback;
        private static int _messagesLogged;

        /// <summary>
        /// Ask the driver to say which call it is refusing.
        ///
        /// GL_INVALID_OPERATION on its own names nothing: it is the error for
        /// several dozen different mistakes, and a frame that raises one every
        /// time can be refusing anything. KHR_debug is the extension that
        /// turns it into a sentence, and it is the difference between another
        /// guess and an answer. Off in normal play -- it costs a callback per
        /// message and there is nobody to read them.
        /// </summary>
        public static void EnableDebugOutput(Action<string> report)
        {
            try
            {
                // OpenTK can call a null native function pointer for an absent
                // extension. A managed catch cannot recover from that fault.
                if (OperatingSystem.IsAndroid()
                    || (ContextVersion() < new Version(4, 3) && !GLFW.ExtensionSupported("GL_KHR_debug"))
                    || GLFW.GetProcAddress("glDebugMessageCallback") == IntPtr.Zero)
                {
                    report("GL debug output unavailable; continuing without optional diagnostics.");
                    return;
                }
                _messagesLogged = 0;
                _debugCallback = (source, type, id, severity, length, message, param) =>
                {
                    if (severity == DebugSeverity.DebugSeverityNotification || _messagesLogged >= 12)
                    {
                        return;
                    }
                    _messagesLogged++;
                    try
                    {
                        string text = System.Runtime.InteropServices.Marshal.PtrToStringAnsi(message, length);
                        report($"GL says: [{severity}] {type} from {source}: {text}");
                    }
                    catch (Exception)
                    {
                        // Diagnostic sinks must never throw through a driver callback.
                    }
                };
                GL.Enable(EnableCap.DebugOutput);
                GL.Enable((EnableCap)All.DebugOutputSynchronous);
                GL.DebugMessageCallback(_debugCallback, IntPtr.Zero);
            }
            catch (Exception ex)
            {
                report($"could not turn on GL debug output ({ex.GetType().Name}); "
                    + "this driver may not have KHR_debug");
            }
        }

        private static Version ContextVersion()
        {
            string first = (GL.GetString(StringName.Version) ?? "").Split(' ')[0];
            return Version.TryParse(first, out var version) ? version : new Version(0, 0);
        }

        public static string DescribeContext()
        {
            try
            {
                string vendor = GL.GetString(StringName.Vendor) ?? "?";
                string renderer = GL.GetString(StringName.Renderer) ?? "?";
                string version = GL.GetString(StringName.Version) ?? "?";
                Version contextVersion = ContextVersion();
                int flags = contextVersion >= new Version(3, 0)
                    ? GL.GetInteger((GetPName)All.ContextFlags) : 0;
                // The one that actually decides whether immediate mode
                // exists. The profile mask can say "compatibility" while this
                // bit has already removed every deprecated entry point.
                string forward = (flags & (int)All.ContextFlagForwardCompatibleBit) != 0
                    ? ", FORWARD-COMPATIBLE (deprecated entry points removed, "
                        + "which is all of immediate mode)"
                    : "";
                int mask = contextVersion >= new Version(3, 2)
                    ? GL.GetInteger((GetPName)All.ContextProfileMask) : 0;
                string profile = contextVersion < new Version(3, 2) ? "legacy"
                    : (mask & (int)All.ContextCoreProfileBit) != 0
                    ? "CORE (immediate mode is unavailable, which renders everything black)"
                    : (mask & (int)All.ContextCompatibilityProfileBit) != 0 ? "compatibility" : "unreported";
                return $"GL {version}, profile {profile}{forward}, {vendor} / {renderer}";
            }
            catch (Exception ex)
            {
                return $"could not query the GL context: {ex.Message}";
            }
        }

        /// <summary>
        /// How much of the image is not the clear colour, as a fraction.
        ///
        /// A capture that is entirely black is the failure this whole path
        /// has to be able to detect: it looks like a successful render in
        /// every log, and only reading the pixels back says otherwise.
        /// </summary>
        public static double NonBlackFraction(Scene scene)
        {
            byte[]? pixels = scene.ReadSceneTarget(out int width, out int height);
            if (pixels == null || width <= 0)
            {
                return 0;
            }
            int lit = 0;
            for (int i = 0; i < pixels.Length; i += 3)
            {
                if (pixels[i] > 8 || pixels[i + 1] > 8 || pixels[i + 2] > 8)
                {
                    lit++;
                }
            }
            return lit / (double)(width * height);
        }
    }
}
